### ISSUE TITLE
Enforce destructuring

### ISSUE THREAD

**Issue by rwjblue** [dockyard/ember-suave#5]
```javascript
// bad
let foo = SomeThing.foo;

// bad
let { foo } = SomeThing;
let { bar } = SomeThing;

// bad
let parts = SomeThing.split(' ');
let first = parts[0];
let second = parts[1];

// bad
let bob = get(SomeThing, 'bob');
let joe = get(SomeThing, 'joe');

// good
let { foo } = SomeThing;
let { foo, bar } = SomeThing;
let { foo, bar } = SomeThing;
let [ first, second ] = SomeThing.split(' ');
let { bob, joe } = getProperties(SomeThing, 'bob', 'joe');

// temp good (due to bug in JSCS/Esprima)
let set = SomeThing.set;
let get = SomeThing.get;
```

**Comment by poteto** [comment]
```js 
// temp good (due to bug in JSCS/Esprima)
let set = SomeThing.set;
let get = SomeThing.get;
```

Another workaround:

```js
let { get: get } = SomeThing;
```

**Comment by rwjblue** [comment]
Yeah, I'm going to test in a vanilla JS + JSCS setup and report an issue to see what needs to be done to fix it.  I'm afraid that the issue is with parsing this as an object literal setter function (ala `set name() { }`), in which case the bug would likely be in Esprima itself...

**Comment by brzpegasus** [comment]
There's a lot of things we check for here. I was thinking of breaking them up into the following rules:

* **requireObjectDestructuring**
```javascript
// bad
let foo = SomeThing.foo;

// good
let { foo } = SomeThing;

// also good for now
let set = SomeThing.set;
let get = SomeThing.get;
```

* **requireArrayDestructuring**
```javascript
// bad
let parts = SomeThing.split(' ');
let first = parts[0];
let second = parts[1];

// good
let [ first, second ] = SomeThing.split(' ');
```

* **requireObjectDestructuringForGet**
```javascript
// bad
let bob = get(SomeThing, 'bob');
let joe = get(SomeThing, 'joe');

// good
let { bob, joe } = getProperties(SomeThing, 'bob', 'joe');
```

Also, just to clarify, are we ok not using object/array destructuring if it's only one assignment? How strict do we want to be? I feel like `let foo = SomeThing.foo;` should be ok so long as we don't have multiple sequential assignments that grab properties from the same object.

**Comment by rwjblue** [comment]
> requireObjectDestructuring

:+1:

> requireObjectDestructuringForGet

I think that the Ember specific rules should include `Ember` in the rule name (to make it clearer).  So the `requireObjectDestructuringForGet` would be something like `requireEmberObjectDestructuringForGet`.

> are we ok not using object/array destructuring if it's only one assignment?

I think that we should likely require destructuring in all cases where the local variable name is the same as the property name off the object. It removes extra line noise and duplication.

**Comment by poteto** [comment]
I'm a little bit unsure about `requireObjectDestructuringForGet` (saying that with the assumption that you haven't decided whether or not to enforce it) – it seems rather verbose and we're repeating the key name twice.

As for single assignments, I think it's better to use destructuring anyway as it keeps things consistent and easy to add new properties.

**Comment by michaeldupuisjr** [comment]
Maybe add multiple lines for multiple assignments? I'm a fan, but I don't know how other people feel about this or if it's important enough to enforce.

```javascript
// good
let {
  firstItem,
  secondItem
} = SomeThing;
```

**Comment by brzpegasus** [comment]
@michaeldupuisjr That can be yet another rule =).

**Comment by miguelcobain** [comment]
I tried to

```js
let [bar1] = this.$('.md-bar1');
```

but got `Invalid attempt to destructure non-iterable instance`.

temporary workaround:

```js
let bar1 = this.$('.md-bar1').get(0);
```

any ideas?

**Comment by rwjblue** [comment]
The returned jQuery object isn't an array, and it doesn't implement the iterable symbol, so it can't be destructured from.

**Comment by miguelcobain** [comment]
Correct, but the jscs rule still enforces me to (wrongly) destructure.

**Comment by brzpegasus** [comment]
This is the alternative syntax we've been using:

```js
let bar1 = this.$('.md-bar1').get(0);
```

If you prefer to do `[0]` instead of `.get(0)`, you could disable the rule for that line with `// jscs:disable`. Unfortunately, it is not trivial (or even possible?) for the rule to detect whether the thing you are calling `[0]

**Comment by hzoo** [comment]
This could possibly be done with a flow/typescript annotation or if we could hook into their system (that's probably a difficult feature to add, although very useful)
